Support mutating constant attributes in export

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110471

In this PR, we implement simulating setattr on nn.Module and HuggingFacePretrainedConfig objets while not changing the original object. For nn.Module, we do:

1. If the nn.Module has custom setattr, we inline it
2. If the nn.Module has default setattr, we inline it in torch.compile and simulate this behavior in export only. The reason is inlining setattr will cause a graph break (https://github.com/pytorch/pytorch/blame/f2d7faf4ba92a6ed43890775ca6ca174ddbf99ea/torch/nn/modules/module.py#L1755) This is fine because this simulation only happens for constant attributes so it is pretty easy to keep track.

For HuggingFacePreTrained, we do:
1. Inline setattr for torch.compile 
2. Simulate the setattr in torch.export. This seems bit problematic for attributes that are stored inside https://github.com/huggingface/transformers/blob/2f3ea08a077ba3133fa8a604b22436cad250b055/src/transformers/configuration_utils.py#L254 but i am not really sure how likely these occur in practice. If the issue rises, we can probably revisit this issue? 

In the future, I think we should create ExportInstructionTranslator that knows how to deal with mutations. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

tugsbayasgalan committed:
Support mutating constant attributes in export

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support mutating constant attributes in export"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

avikchaudhuri commented on test/dynamo/test_export_mutations.py:
```diff
@@ -13,7 +14,8 @@ def check_failure_on_export(self, mod, *args):
 
     def check_same_with_export(self, mod, arg):
-        real_result = mod(arg)
+        eager_mod = copy.deepcopy(mod)
```
Is deep copying usually how such things are tested? IMO it's better to take the module class instead, and just create the module instance twice.

tugsbayasgalan replied:
Yeah good idea I can do that

avikchaudhuri commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2304,7 +2304,14 @@ def inline_call_(
+                (
+                    variables.CustomizedDictVariable,
+                    variables.NNModuleVariable,
+                    variables.dicts.HFPretrainedConfigVariable,
```
Curious: why is this so special? There are a bunch of others in `variables.dicts` as well...

avikchaudhuri replied:
Looks like a sentinel for this could be `constant_attributes_tracker_EXPORT_ONLY`. In general I'd like these to share some kind of protocol instead of listing them manually here.

tugsbayasgalan replied:
I actually don't think this is needed.

avikchaudhuri reviewed (commented):
I don't feel confident in my review, left some nits. @anijain2305 should take a look...

avikchaudhuri commented on test/dynamo/test_export_mutations.py:
```diff
@@ -128,6 +92,44 @@ def forward(self, x):
+
+        self.check_same_with_export(Foo(), torch.Tensor(3, 2))
+
+    def test_module_attribute_mutation_violation_negative_7(self):
```
while we're at it, I'd really love if we can properly name these tests instead of using numbers. All of these share `module_attribute_mutation_violation` so you can simply shorten the names to begin with, and then use some descriptive suffixes (maybe keeping positive / negative, because that's kind of useful...or, if you wish, success / failure).

tugsbayasgalan replied:
Hahah sounds good

tugsbayasgalan committed:
Update on "Support mutating constant attributes in export"


In this PR, we implement simulating setattr on nn.Module and HuggingFacePretrainedConfig objets while not changing the original object. For nn.Module, we do:

1. If the nn.Module has custom setattr, we inline it
2. If the nn.Module has default setattr, we inline it in torch.compile and simulate this behavior in export only. The reason is inlining setattr will cause a graph break (https://github.com/pytorch/pytorch/blame/f2d7faf4ba92a6ed43890775ca6ca174ddbf99ea/torch/nn/modules/module.py#L1755) This is fine because this simulation only happens for constant attributes so it is pretty easy to keep track.

For HuggingFacePreTrained, we do:
1. Inline setattr for torch.compile 
2. Simulate the setattr in torch.export. This seems bit problematic for attributes that are stored inside https://github.com/huggingface/transformers/blob/2f3ea08a077ba3133fa8a604b22436cad250b055/src/transformers/configuration_utils.py#L254 but i am not really sure how likely these occur in practice. If the issue rises, we can probably revisit this issue? 

In the future, I think we should create ExportInstructionTranslator that knows how to deal with mutations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng 

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support mutating constant attributes in export"


In this PR, we implement simulating setattr on nn.Module and HuggingFacePretrainedConfig objets while not changing the original object. For nn.Module, we do:

1. If the nn.Module has custom setattr, we inline it
2. If the nn.Module has default setattr, we inline it in torch.compile and simulate this behavior in export only. The reason is inlining setattr will cause a graph break (https://github.com/pytorch/pytorch/blame/f2d7faf4ba92a6ed43890775ca6ca174ddbf99ea/torch/nn/modules/module.py#L1755) This is fine because this simulation only happens for constant attributes so it is pretty easy to keep track.

For HuggingFacePreTrained, we do:
1. Inline setattr for torch.compile 
2. Simulate the setattr in torch.export. This seems bit problematic for attributes that are stored inside https://github.com/huggingface/transformers/blob/2f3ea08a077ba3133fa8a604b22436cad250b055/src/transformers/configuration_utils.py#L254 but i am not really sure how likely these occur in practice. If the issue rises, we can probably revisit this issue? 

In the future, I think we should create ExportInstructionTranslator that knows how to deal with mutations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng 

[ghstack-poisoned]
